In [25]:
import pandas as pd
excel_data = pd.read_excel('C:/Users/msi/Downloads/DATA EUROMILLION CHOIX DISCRET (1).xlsx')


### Les varaibles exogéne

JOUR tirage: Cette colonne indique le jour de la semaine où le tirage de l'EuroMillions a eu lieu. Les valeurs typiques pourraient être "Lundi", "Mercredi", "Vendredi", etc., en fonction des jours de tirage pour l'EuroMillions.

E1 et E2: Ces colonnes représentent les numéros des deux étoiles tirées dans l'EuroMillions. Les tirages de l'EuroMillions comprennent généralement la sélection de deux numéros supplémentaires, appelés étoiles, en plus des numéros réguliers.

pair et impair: Ces colonnes semblent indiquer le nombre de numéros pairs et impairs tirés dans chaque jeu. Par exemple, si dans un tirage particulier, trois numéros pairs et deux numéros impairs sont tirés, les valeurs correspondantes dans ces colonnes seraient 3 pour "pair" et 2 pour "impair".

pair_E1: Cette colonne pourrait indiquer si le premier numéro étoile (E1) est pair. Elle prendrait probablement une valeur de 1 si E1 est pair, et 0 si E1 est impair.

sum_b: Cette colonne représente probablement la somme des numéros des boules tirées (B1, B2, B3, B4, B5). Cela donne un aperçu de la distribution des numéros dans un tirage particulier.

is_sumpaire: Cette colonne semble indiquer si la somme des numéros de boules (sum_b) est un nombre pair. Si la somme est paire, la valeur serait 1; si elle est impaire, la valeur serait 0.

corr: Cette colonne pourrait représenter une mesure de corrélation, bien que le contexte spécifique ne soit pas clair sans informations supplémentaires. Elle pourrait indiquer une sorte de corrélation statistique entre différents éléments du tirage, comme les numéros de boules, les numéros d'étoiles, ou d'autres facteurs.

### Les varaibles endogéne

Pour déterminer la variable endogène \( Y \), les critères suivants sont utilisés en fonction de la somme des numéros des boules tirées :

- **Si la somme est inférieure à 50** : \( Y = 0 \).
- **Si la somme est entre 50 et 100 (inclus)** : \( Y = 1 \).
- **Si la somme est entre 100 et 150 (inclus)** : \( Y = 2 \).
- **Si la somme est entre 150 et 200 (inclus)** : \( Y = 3 \).
- **Si la somme est entre 200 et 250 (inclus)** : \( Y = 4 \).

Chaque intervalle de somme correspond à une modalité spécifique de \( Y \).
 

In [26]:
# les varaibles exogéne utilisée
liste_columns=['JOUR tirage','E1','E2','pair','impair','pair_E1','sum_b','is_sumpaire','corr']

In [27]:
# Importation des bibliothèques nécessaires
import pandas as pd
import statsmodels.api as sm


# Fonction pour catégoriser la varaible endogéne
def categorize_and_count(row):
    count = 0
    for number in row:
        count = number + count
    if count <=50:
      return 0
    if  50 < count <= 100:
      return 1 
    if  100 < count <= 150:
          return 2 
    if  150 < count <= 200:
          return 3 
    else: 
        return 4


### Dans cette section nous avons choisi le trige de mardi 

In [28]:
excel_data_mardi= excel_data[excel_data["JOUR tirage"]=="MARDI   "]

In [29]:
# Application de la fonction aux colonnes B1 à B5 pour créer la variable endogène
excel_data_mardi["Endogéne"] = excel_data_mardi[['B1', 'B2', 'B3', 'B4', 'B5']].apply( lambda row: categorize_and_count(row), axis=1)

C:\Users\msi\AppData\Local\Temp\ipykernel_43316\3863096803.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  excel_data_mardi["Endogéne"] = excel_data_mardi[['B1', 'B2', 'B3', 'B4', 'B5']].apply( lambda row: categorize_and_count(row), axis=1)


Nous avons réussi à coder la variable endogène en utilisant cinq modalités distinctes, allant de 0 à 4. Parallèlement, nous avons entamé le processus de codage des variables exogènes catégorielles sous une forme numérique. Cette étape implique la transformation des attributs catégoriels, qui sont souvent sous forme de labels ou de classes, en valeurs numériques. Cette conversion est cruciale pour l'analyse statistique et les modèles prédictifs, car elle permet d'intégrer ces variables dans des algorithmes qui requièrent des données numériques.

Parmi les variables exogènes sélectionnées pour notre analyse, nous avons deux catégories principales : les numéros d'étoiles gagnants et le jour de tirage. Pour convertir ces données en format numérique, nous avons élaboré la stratégie suivante :

Conversion des numéros d'étoiles : Nous avons décidé que si la somme des numéros d'étoiles est inférieure à 0, cette somme sera codée comme 10. Dans le cas contraire, la somme sera codée comme 1. Cette approche simplifie la représentation des numéros d'étoiles en les regroupant en deux catégories distinctes basées sur leur somme totale.

Conversion du jour de tirage : Pour cette variable, nous avons envisagé plusieurs options :

Option 1 : Modèle global avec deux modalités pour les jours : Dans cette méthode, nous transformons les jours de tirage en deux catégories. Si le tirage a lieu un vendredi, nous le codons comme 0. Pour tous les autres jours, nous le codons comme 1. Cette approche nous permet de distinguer le vendredi des autres jours, en supposant que le vendredi pourrait avoir une caractéristique spécifique influençant les tirages.
Option 2 : Analyse spécifique pour chaque jour : Une autre stratégie consiste à analyser les données en fonction de chaque jour de tirage séparément. Cela implique de créer des modèles distincts pour chaque jour, permettant une compréhension plus détaillée des tendances et des schémas qui peuvent varier selon le jour de la semaine.
Ces méthodes de codage des variables catégorielles en formats numériques sont essentielles pour intégrer ces informations dans des modèles statistiques et des analyses prédictives. Elles nous permettent de simplifier les caractéristiques catégorielles tout en préservant les informations clés nécessaires pour nos analyses.

In [30]:
def categorize_étoile(row):
    count = 0
    for number in row:
        count = number + count
    if count <=10:
      return 0 
    else: 
        return 2

In [31]:
excel_data_mardi['Groupe_E1_E2']= excel_data_mardi[['E1', 'E2']].apply( lambda row: categorize_étoile(row), axis=1)

C:\Users\msi\AppData\Local\Temp\ipykernel_43316\39110240.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  excel_data_mardi['Groupe_E1_E2']= excel_data_mardi[['E1', 'E2']].apply( lambda row: categorize_étoile(row), axis=1)


In [32]:
# Sélection de quelques variables exogènes pour le modèle
Exogéne_varaibles = excel_data_mardi[['Groupe_E1_E2','pair','impair','pair_E1','sum_b','is_sumpaire','corr','MAX','MIN','produit']]
endogenous_variable = excel_data_mardi["Endogéne"]

In [33]:
# Ajout d'une colonne constante aux variables exogènes pour le terme d'interception
X = sm.add_constant(Exogéne_varaibles)


In [34]:

# Création et ajustement du modèle de régression de Poisson avec statsmodels
poisson_model = sm.GLM(endogenous_variable, X, family=sm.families.Poisson())
poisson_results = poisson_model.fit()

# Affichage du résumé du modèle
poisson_results_summary = poisson_results.summary()

In [35]:
# Affichage du résumé du modèle
poisson_results_summary 

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:               Endogéne   No. Observations:                  654
Model:                            GLM   Df Residuals:                      644
Model Family:                 Poisson   Df Model:                            9
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -856.67
Date:                Sun, 07 Jan 2024   Deviance:                       29.430
Time:                        16:21:46   Pearson chi2:                     28.8
No. Iterations:                     4   Pseudo R-squ. (CS):             0.1627
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.0666      0.031     -2.154      0.031      -0.127      -0.006
Groupe_E1_E2    -0.0071      0.029     -0.250      0.803      -0.063       0.049
pair            -0.1677      0.080     -2.103      0.035      -0.324      -0.011
impair          -0.1655      0.077     -2.150      0.032      -0.316      -0.015
pair_E1          0.0049      0.055      0.089      0.929      -0.103       0.113
sum_b            0.0109      0.002      5.391      0.000       0.007       0.015
is_sumpaire     -0.0312      0.055     -0.563      0.573      -0.140       0.077
corr             0.0933      0.409      0.228      0.820      -0.708       0.895
MAX              0.0027      0.007      0.403      0.687      -0.010       0.016
MIN              0.0045      0.007      0.668      0.504      -0.009       0.018
produit      -4.181e-09   3.27e-09     -1.280      0.201   -1.06e-08    2.22e-09
================================================================================
"""

Résultats du Modèle:

Variable Dépendante (Dep. Variable) : y

Nombre d'Observations (No. Observations) : 654

Degrés de liberté résiduels (Df Residuals) : 644

Famille du Modèle (Model Family) : Poisson

Degrés de liberté du modèle (Df Model) : 9

Fonction de lien (Link Function) : Log

Méthode (Method) : IRLS

Log-vraisemblance (Log-Likelihood) : -856.67

Déviance (Deviance) : 29.430

Chi2 de Pearson (Pearson chi2) : 28.8

Nombre d'itérations (No. Iterations) : 4


Pseudo R-carré (Pseudo R-squ.) : 0.1627



Analyse des Résultats

Constante (const):
Coefficient : -0.0666
P-value : 0.031
La constante est statistiquement significative (p < 0.05). Cela indique une influence notable sur la variable dépendante lorsque toutes les autres variables sont à zéro.

Groupe_E1_E2:
Coefficient : -0.0071
P-value : 0.803
Non significatif statistiquement. Suggère peu ou pas d'impact sur la variable dépendante.

pair:
Coefficient : -0.1677
P-value : 0.035
Statistiquement significatif. Indique une influence négative sur la variable dépendante.

impair:
Coefficient : -0.1655
P-value : 0.032
Statistiquement significatif, similaire à pair, indiquant également une influence négative.

pair_E1:
Coefficient : 0.0049
P-value : 0.929
Non significatif statistiquement.

sum_b:
Coefficient : 0.0109
P-value : <0.001
Très significatif statistiquement. Forte influence positive sur la variable dépendante.

is_sumpaire:
Coefficient : -0.0312
P-value : 0.573
Non significatif statistiquement.

corr:
Coefficient : 0.0933
P-value : 0.820
Non significatif statistiquement.

MAX:
Coefficient : 0.0027
P-value : 0.687
Non significatif statistiquement.

MIN:
Coefficient : 0.0045
P-value : 0.504
Non significatif statistiquement.


produit:
Coefficient : -4.181e-09
P-value : 0.201
Non significatif statistiquement.


Recommandations sur les Variables

À Conserver:
sum_b : Très significative, indique une forte influence.
pair et impair : Significatives, indiquent des influences négatives.
À Écarter :

Groupe_E1_E2, pair_E1, is_sumpaire, corr, MAX, MIN, produit : Ces variables ne sont pas statistiquement significatives et peuvent être écartées pour simplifier le modèle.

-Interprétation Générale :

Le modèle semble bien ajusté avec un nombre d'itérations raisonnable et un pseudo R-carré de 0.1627, indiquant une bonne adéquation du modèle avec les données. La déviance et le Chi2 de Pearson sont également relativement bas, ce qui est positif pour le modèle.

Dans cette section on va garder que les varaibles qui ont une importance

In [36]:
# Sélection de quelques variables exogènes pour le modèle
Exogéne_varaibles = excel_data_mardi[['pair','impair','sum_b']]

In [37]:
# Ajout d'une colonne constante aux variables exogènes pour le terme d'interception
X = sm.add_constant(Exogéne_varaibles)

In [38]:
# Création et ajustement du modèle de régression de Poisson avec statsmodels
poisson_model = sm.GLM(endogenous_variable, X, family=sm.families.Poisson())
poisson_results = poisson_model.fit()

# Affichage du résumé du modèle
poisson_results_summary = poisson_results.summary()

In [39]:
poisson_results_summary

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:               Endogéne   No. Observations:                  654
Model:                            GLM   Df Residuals:                      651
Model Family:                 Poisson   Df Model:                            2
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -858.02
Date:                Sun, 07 Jan 2024   Deviance:                       32.132
Time:                        16:21:46   Pearson chi2:                     31.1
No. Iterations:                     4   Pseudo R-squ. (CS):             0.1593
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0455      0.010     -4.647      0.000      -0.065      -0.026
pair          -0.1133      0.028     -4.068      0.000      -0.168      -0.059
impair        -0.1141      0.027     -4.187      0.000      -0.168      -0.061
sum_b          0.0100      0.001     10.546      0.000       0.008       0.012
==============================================================================
"""

Le modèle de régression de Poisson appliqué sur 654 observations révèle des résultats significatifs. Avec un log-vraisemblance de -858.02 et un pseudo R-carré de 0.1593, le modèle montre une bonne adéquation. Les coefficients pour pair (-0.1133) et impair (-0.1141) sont tous deux significativement négatifs, suggérant une influence négative sur la variable dépendante y. En revanche, sum_b a un effet positif notable (coefficient de 0.0100), indiquant une augmentation de y avec une augmentation de sum_b. Le modèle, avec un nombre d'itérations de 4 et une déviance de 32.132, semble bien ajusté aux données. Ces résultats indiquent que les facteurs considérés ont un impact significatif sur la variable dépendante dans ce contexte statistique spécifique.

In [40]:
# Calcul des effets marginaux
marginal_effects = poisson_results.get_margeff()
print(marginal_effects.summary())

         GLM Marginal Effects        
Dep. Variable:               Endogéne
Method:                          dydx
At:                           overall
                dy/dx    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
pair          -0.2296      0.056     -4.127      0.000      -0.339      -0.121
impair        -0.2312      0.054     -4.250      0.000      -0.338      -0.125
sum_b          0.0203      0.002     10.129      0.000       0.016       0.024


Dans le modèle GLM Poisson, les effets marginaux des variables pair, impair et sum_b sur la variable dépendante y sont significatifs et instructifs. Pour pair, l'effet marginal est de -0.2296, indiquant qu'une augmentation d'une unité dans pair entraîne une baisse de 0.2296 dans la valeur attendue de y. Cette influence négative est significative, avec une valeur de p proche de zéro. De manière similaire, impair a un effet marginal de -0.2312, montrant aussi un impact négatif substantiel sur y. En revanche, sum_b présente un effet positif avec un effet marginal de 0.0203, ce qui signifie que chaque augmentation d'une unité dans sum_b augmente la valeur attendue de y de 0.0203. Ces résultats sont statistiquement significatifs, soulignant l'influence notable de ces variables sur la variable dépendante dans notre modèle.

In [41]:
X.describe()

const        pair      impair       sum_b
count  654.0  654.000000  654.000000  654.000000
mean     1.0    2.475535    2.524465  127.169725
std      0.0    1.088087    1.088087   29.636443
min      1.0    0.000000    0.000000   44.000000
25%      1.0    2.000000    2.000000  106.000000
50%      1.0    2.000000    3.000000  128.500000
75%      1.0    3.000000    3.000000  148.000000
max      1.0    5.000000    5.000000  222.000000

In [42]:
import pandas as pd

# Création de la nouvelle observation avec les valeurs fournies
new_observation = {
    'const': [ 1, 1, 1, 1, 1],
    'pair': [0, 1, 2, 3, 4],
    'impair': [4, 3, 2, 1, 0],
    'sum_b': [127.169725, 127.169725, 127.169725, 127.169725, 127.169725]
}

# Conversion en DataFrame
df_new = pd.DataFrame.from_dict(new_observation)
df_new

const  pair  impair       sum_b
0      1     0       4  127.169725
1      1     1       3  127.169725
2      1     2       2  127.169725
3      1     3       1  127.169725
4      1     4       0  127.169725

In [43]:
# Si df_new n'a pas la colonne 'const', ajoutez-la
if 'const' not in df_new.columns:
    df_new = sm.add_constant(df_new)

# Faire la prédiction en utilisant le modèle
predicted_nombre_point= poisson_results.predict(df_new)
prediction_lambda = list(predicted_nombre_point)

# Création de la nouvelle observation avec les valeurs fournies
new_observation = {
    'const': [1, 1, 1, 1, 1],
    'pair': [0, 1, 2, 3, 4],
    'impair': [4, 3, 2, 1, 0],
    'sum_b': [127.169725, 127.169725, 127.169725, 127.169725, 127.169725],
    'estimation_lambda': prediction_lambda
}

# Conversion en DataFrame
df_new = pd.DataFrame.from_dict(new_observation)
df_new

const  pair  impair       sum_b  estimation_lambda
0      1     0       4  127.169725           2.169930
1      1     1       3  127.169725           2.171608
2      1     2       2  127.169725           2.173287
3      1     3       1  127.169725           2.174968
4      1     4       0  127.169725           2.176650